File: py800mon/shortcuts.py
import curses

from .ui import Color


class ShortcutLayerNotFound(Exception):
    pass


class ShortcutAlreadyRegistered(Exception):
    pass


class LayerAlreadyRegistered(Exception):
    pass


KEYS_STR_MAP = {
    27: "Esc",
    9: "Tab",
}


def _normalize_key(key: int) -> int:
    v = int(key)
    if ord("A") <= v <= ord("Z"):
        return v + 32
    return v


def key_as_text(key: int) -> str:
    key = _normalize_key(key)
    if key in KEYS_STR_MAP:
        return KEYS_STR_MAP[key]

    if key >= curses.KEY_F0 and key <= curses.KEY_F0 + 63:
        return f"F{key - curses.KEY_F0}"

    if key < 32:
        return "^" + chr(key + 64)

    if key > 126:
        return str(key)

    return chr(key).upper()


class Shortcut:
    def __init__(
        self,
        key,
        label: str,
        callback: callable,
        visible_in_global_bar: bool = True,
    ):
        if isinstance(key, str):
            self.key = _normalize_key(ord(key[0]))
        else:
            self.key = _normalize_key(int(key))
        self.label = label
        self.callback = callback
        self.visible_in_global_bar = visible_in_global_bar

    def key_as_text(self):
        return key_as_text(self.key)

    def __repr__(self):
        return f'<Shortcut key="{self.key_as_text()}" label="{self.label}">'


class ShortcutLayer:
    def __init__(self, name="", color: Color = Color.APPMODE):
        self.name = name
        self.color = color
        self._shortcuts = {}

    def add(self, shortcut: Shortcut):
        if shortcut.key in self._shortcuts:
            raise ShortcutAlreadyRegistered(shortcut)
        self._shortcuts[shortcut.key] = shortcut

    def get_shortcuts(self):
        return self._shortcuts.values()

    def get(self, key):
        return self._shortcuts[_normalize_key(key)]

    def has(self, key):
        return _normalize_key(key) in self._shortcuts


class ShortcutManager:
    def __init__(self, initial=None):
        self._globals = {}
        self._layers = {}

    def get(self, layer: str):
        return self._layers.get(layer)

    def add_global(self, shortcut: Shortcut):
        if shortcut.key in self._globals:
            raise ShortcutAlreadyRegistered(shortcut)
        self._globals[shortcut.key] = shortcut

    def has_global(self, key: int):
        return _normalize_key(key) in self._globals

    def get_global(self, key: int):
        return self._globals[_normalize_key(key)]

    def global_shortcuts(self):
        return [s for s in self._globals.values() if s.visible_in_global_bar]

    def add(self, layer_id: str, layer: ShortcutLayer):
        if layer_id in self._layers:
            raise LayerAlreadyRegistered(layer_id)
        self._layers[layer_id] = layer

File: go800mon/a800mon/shortcuts.go
package a800mon

import "fmt"

type Shortcut struct {
	Key                int
	Label              string
	Callback           func()
	VisibleInGlobalBar bool
}

func NewShortcut(key int, label string, callback func()) Shortcut {
	return Shortcut{
		Key:                normalizeShortcutKey(key),
		Label:              label,
		Callback:           callback,
		VisibleInGlobalBar: true,
	}
}

func (s Shortcut) KeyAsText() string {
	return shortcutKeyAsText(s.Key)
}

type ShortcutLayer struct {
	Name      string
	Color     Color
	shortcuts map[int]Shortcut
	order     []int
}

func NewShortcutLayer(name string, color Color) *ShortcutLayer {
	return &ShortcutLayer{Name: name, Color: color, shortcuts: map[int]Shortcut{}, order: []int{}}
}

func (l *ShortcutLayer) Add(shortcut Shortcut) error {
	key := normalizeShortcutKey(shortcut.Key)
	shortcut.Key = key
	if _, ok := l.shortcuts[key]; ok {
		return fmt.Errorf("shortcut already registered: %d", key)
	}
	l.shortcuts[key] = shortcut
	l.order = append(l.order, key)
	return nil
}

func (l *ShortcutLayer) Get(key int) (Shortcut, bool) {
	s, ok := l.shortcuts[normalizeShortcutKey(key)]
	return s, ok
}

func (l *ShortcutLayer) Has(key int) bool {
	_, ok := l.shortcuts[normalizeShortcutKey(key)]
	return ok
}

func (l *ShortcutLayer) List() []Shortcut {
	out := make([]Shortcut, 0, len(l.order))
	for _, key := range l.order {
		out = append(out, l.shortcuts[key])
	}
	return out
}

type ShortcutManager struct {
	globals      map[int]Shortcut
	globalsOrder []int
	layers       map[int]*ShortcutLayer
}

func NewShortcutManager() *ShortcutManager {
	return &ShortcutManager{
		globals:      map[int]Shortcut{},
		globalsOrder: []int{},
		layers:       map[int]*ShortcutLayer{},
	}
}

func (m *ShortcutManager) AddGlobal(shortcut Shortcut) error {
	key := normalizeShortcutKey(shortcut.Key)
	shortcut.Key = key
	if _, ok := m.globals[key]; ok {
		return fmt.Errorf("shortcut already registered: %d", key)
	}
	m.globals[key] = shortcut
	m.globalsOrder = append(m.globalsOrder, key)
	return nil
}

func (m *ShortcutManager) Add(mode int, layer *ShortcutLayer) error {
	if _, ok := m.layers[mode]; ok {
		return fmt.Errorf("layer already registered: %d", mode)
	}
	m.layers[mode] = layer
	return nil
}

func (m *ShortcutManager) Get(mode int) *ShortcutLayer {
	return m.layers[mode]
}

func (m *ShortcutManager) Global(key int) (Shortcut, bool) {
	s, ok := m.globals[normalizeShortcutKey(key)]
	return s, ok
}

func (m *ShortcutManager) HandleInput(mode, key int) bool {
	layer := m.Get(mode)
	if layer != nil {
		if shortcut, ok := layer.Get(key); ok {
			if shortcut.Callback != nil {
				shortcut.Callback()
			}
			return true
		}
	}
	if shortcut, ok := m.Global(key); ok {
		if shortcut.Callback != nil {
			shortcut.Callback()
		}
		return true
	}
	return false
}

func normalizeShortcutKey(key int) int {
	if key >= int('A') && key <= int('Z') {
		return key + 32
	}
	return key
}

func shortcutKeyAsText(key int) string {
	key = normalizeShortcutKey(key)
	if key == 27 {
		return "Esc"
	}
	if key == 9 {
		return "Tab"
	}
	if key >= KeyF0() && key <= KeyF0()+63 {
		return fmt.Sprintf("F%d", key-KeyF0())
	}
	if key < 32 {
		return "^" + string(rune(key+64))
	}
	if key > 126 {
		return fmt.Sprintf("%d", key)
	}
	if key >= int('a') && key <= int('z') {
		return string(rune(key - 32))
	}
	return string(rune(key))
}

func (m *ShortcutManager) Globals() []Shortcut {
	out := make([]Shortcut, 0, len(m.globalsOrder))
	for _, key := range m.globalsOrder {
		shortcut := m.globals[key]
		if !shortcut.VisibleInGlobalBar {
			continue
		}
		out = append(out, shortcut)
	}
	return out
}
